tests/FunctionalTestBase.php

```diff
@@ -0,0 +1,125 @@
+        // with the internet.
+        $lock = __DIR__ . '/../composer.lock';
+        static::assertFileIsReadable($lock);
+        $lock = file_get_contents($lock);
+        $lock = json_decode($lock, true);
+        $vendor = [];
+        $packages = array_merge($lock['packages'], $lock['packages-dev']);
+        foreach ($packages as $package) {
+            $name = $package['name'];
+            $dir = __DIR__ . '/../vendor/' . $name;
+            if (is_dir($dir)) {
+                $version = $package['version'];
+                $vendor['packages'][$name][$version] = [
+                    'name' => $name,
+                    'version' => $version,
+                    'type' => $package['type'],
+                    'dist' => [
+                        'type' => 'path',
+                        'url' => $dir,
+                    ],
+                ];
+            }
+        }
+        $destination = $this->workingDir . '/vendor.json';
+        file_put_contents($destination, json_encode($vendor, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES));
+        $this->composer('config', 'repo.vendor', 'composer', 'file://' . $destination);
+    }
+
+    protected function startServer(): void
+    {
+        $this->server = new Process([PHP_BINARY, '-S', 'localhost:8080'], __DIR__ . '/_targets');
+        $this->server->start();
+        $serverStarted = $this->server->waitUntil(function ($outputType, $output): bool {
+            return str_contains($output, 'Development Server (http://localhost:8080) started');
+        });
+        static::assertTrue($serverStarted);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    protected function tearDown(): void
+    {
+        // Stop the web server.
+        $this->server?->stop();
+
+        // Delete the fake project created for the test.
+        (new Filesystem())->remove($this->workingDir);
+
+        parent::tearDown();
+    }
+
+    /**
+     * Runs Composer in the test project directory.
+     *
+     * @param string ...$arguments
+     *   The arguments to pass to Composer.
+     *
+     * @return Process
+     *   The process object.
+     */
+    protected function composer(string ...$arguments): Process
+    {
+        // Ensure the current PHP runtime is used to execute Composer.
+        array_unshift($arguments, PHP_BINARY, __DIR__ . '/../vendor/composer/composer/bin/composer');
+        // Always run in very, very verbose mode.
+        $arguments[] = '-vvv';
+
+        $process = (new Process($arguments))
+            ->setWorkingDirectory($this->workingDir)
+            ->mustRun();
+        static::assertSame(0, $process->getExitCode());
+        // There should not be any deprecation warnings.
+        static::assertStringNotContainsStringIgnoringCase('deprecated', $process->getOutput());
+        static::assertStringNotContainsStringIgnoringCase('deprecated', $process->getErrorOutput());
+
+        return $process;
```

Comment:
why bother returning this. We never use it, related we should `@internal` this file

Reply:
Yeah, good point. Fixed!